src/Symfony/Bridge/Doctrine/Security/MagicLink/MagicLoginLinkTokenRepositoryTrait.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bridge\Doctrine\Security\MagicLink;
+
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Http\MagicLink\StoredMagicLinkTokenInterface;
+
+/**
+ * Trait that can be used on a Doctrine repository to help implement MagicLinkTokenStorageInterface.
+ *
+ * Using this trait requires following a few conventions:
+ *
+ *      A) Your repository needs a createMagicToken method that returns
+ *          a new instance of your entity:
+ *
+ *          private function createMagicToken(string $selector, string $hashedVerifier, UserInterface $user, \DateTimeInterface $expiresAt): StoredMagicLinkTokenInterface
+ *
+ *      B) Your entity needs a "selector" and "expiresAt" properties.
+ *
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ */
+trait MagicLoginLinkTokenRepositoryTrait
+{
+    public function storeToken(string $selector, string $hashedVerifier, UserInterface $user, \DateTimeInterface $expiresAt): void
+    {
+        if (!method_exists($this, 'createMagicToken')) {
+            throw new \LogicException(sprintf('In order to use MagicLinkRepositoryTrait, the class "%s" must either have a createMagicToken() or override the storeToken() method.', static::class));
+        }
+
+        $magicLoginToken = $this->createMagicToken($selector, $hashedVerifier, $user, $expiresAt);
+
+        $this->getEntityManager()->persist($magicLoginToken);
```

Comment:
shouldn't we check if `getEntityManager()` method exists? (like for `createMagicToken`)

Reply:
I'm not sure - maybe? Technically yes, but there is a "soft contract" that the trait is meant to be used in an entity repository. I'll defer to what others think.